### PR TITLE
Apply new grid to existing layer

### LINKED ISSUE #2283: Regrid functionality
We need a good method to take an already indexed layer (e.g. `TileLayerRDD[SpatialKey]`), and regrid it to a different spatial layout. For instance, I had a use case where I wanted to take a layer that had 512x512 tiles and move it to a layer that had 1024x1024 tiles. I had to hand roll a solution, and we should have library functionality for it. I did a first pass at generalizing this; it needs to account for possible non-square tiles, I currently have both tileCols and tileRows being moved to some single tileSize. Otherwise this methods seems to work; so the work here is to finish generalizing it, placing it in the library and testing it.

```scala
  def regrid[
    K: SpatialComponent: ClassTag,
    V <: CellGrid: ClassTag: Stitcher: (? => CropMethods[V]): (? => TilePrototypeMethods[V]),
    M: Component[?, LayoutDefinition]: Component[?, Bounds[K]]
  ](layer: RDD[(K, V)] with Metadata[M], tileSize: Int): RDD[(K, V)] with Metadata[M] = {
    // println(layer.metadata.toJson.prettyPrint)
    val md = layer.metadata
    val ld = md.getComponent[LayoutDefinition]

    if(ld.tileLayout.tileCols == tileSize) {
      layer
    } else {
      val ceil = { x: Double => math.ceil(x).toInt }

      val tl = ld.tileLayout
      val oldTileSize = tl.tileCols
      if(tileSize > oldTileSize) {
        val r = tileSize / oldTileSize
        val ntl = TileLayout(ceil(tl.layoutCols / r.toDouble), ceil(tl.layoutRows / r.toDouble), tileSize, tileSize)
        val layoutDefinition =
          LayoutDefinition(
            ld.extent,
            ntl
          )

        val bounds =
          md.getComponent[Bounds[K]] match {
            case KeyBounds(minKey, maxKey) =>
              val newMinKey = {
                val SpatialKey(minKeyCol, minKeyRow) = minKey.getComponent[SpatialKey]
                val sk = SpatialKey(minKeyCol / r, minKeyRow / r)
                minKey.setComponent[SpatialKey](sk)
              }
              val newMaxKey = {
                val SpatialKey(maxKeyCol, maxKeyRow) = maxKey.getComponent[SpatialKey]
                val sk = SpatialKey(ceil(maxKeyCol / r.toDouble), ceil(maxKeyRow / r.toDouble))
                maxKey.setComponent[SpatialKey](sk)
              }
              KeyBounds(newMinKey, newMaxKey)
            case EmptyBounds =>
              EmptyBounds
          }

        val newMd =
          md.setComponent[LayoutDefinition](layoutDefinition)
            .setComponent[Bounds[K]](bounds)

        val tiled =
          layer
            .map { case (key, tile) =>
              val sk = key.getComponent[SpatialKey]
              val newKey = key.setComponent[SpatialKey](SpatialKey(sk.col / r, sk.row / r))
              (newKey, (key, tile))
            }
            .groupByKey
            .map { case (key, tiles) =>
              val pieces =
                tiles.map { case (oldKey, oldTile) =>
                  val SpatialKey(oldCol, oldRow) = oldKey.getComponent[SpatialKey]
                  val updateCol = (oldCol % r) * oldTileSize
                  val updateRow = (oldRow % r) * oldTileSize
                  (oldTile, (updateCol, updateRow))
                }
              val newTile =
                implicitly[Stitcher[V]].stitch(pieces, tileSize, tileSize)

              (key, newTile)
          }

        ContextRDD(tiled, newMd)
      } else {
        val r = oldTileSize / tileSize
        val ntl = TileLayout(tl.layoutCols * r, tl.layoutRows * r, tileSize, tileSize)
        val layoutDefinition =
          LayoutDefinition(
            ld.extent,
            ntl
          )

        val bounds =
          md.getComponent[Bounds[K]] match {
            case KeyBounds(minKey, maxKey) =>
              val newMinKey = {
                val SpatialKey(minKeyCol, minKeyRow) = minKey.getComponent[SpatialKey]
                val sk = SpatialKey(minKeyCol * r, minKeyRow * r)
                minKey.setComponent[SpatialKey](sk)
              }
              val newMaxKey = {
                val SpatialKey(maxKeyCol, maxKeyRow) = maxKey.getComponent[SpatialKey]
                val sk = SpatialKey((maxKeyCol + 1) * r - 1, (maxKeyRow + 1) * r - 1)
                maxKey.setComponent[SpatialKey](sk)
              }
              KeyBounds(newMinKey, newMaxKey)
            case EmptyBounds =>
              EmptyBounds
          }

        val newMd =
          md.setComponent[LayoutDefinition](layoutDefinition)
            .setComponent[Bounds[K]](bounds)

        val layoutCols = r
        val layoutRows = r

        val tiled =
          layer.flatMap { case (key, tile) =>
            val result = mutable.ListBuffer[(K, V)]()
            val thisKey = key.getComponent[SpatialKey]
            for(layoutRow <- 0 until layoutRows) {
              for(layoutCol <- 0 until layoutCols) {
                val sk = SpatialKey(thisKey.col * layoutCols + layoutCol, thisKey.row * layoutRows + layoutRow)
                val newTile =
                  tile.crop(
                    GridBounds(
                      layoutCol * tileSize,
                      layoutRow * tileSize,
                      (layoutCol + 1) * tileSize - 1,
                      (layoutRow + 1) * tileSize - 1
                    )
                  )

                result += ((key.setComponent[SpatialKey](sk), newTile))
              }
            }

            result
          }

        ContextRDD(tiled, newMd)
      }
    }
  }
```

### PR DESCRIPTION
**Pull request by jpolchlo**
Fixes #2283 

When presented with a layer containing Tiles of a set resolution, it may be desirable to generate a layer with the same pixel data in a different tile resolution.  This will produce a layer over the same extent with the same cellSize but different KeyBounds.  The exact area of tile coverage may differ (i.e., regridding a 10X10 layout of 32x32 tiles to 100x100 tiles will yield a 4x4 layout, implying a 400x400 overall size—though the tiles on the right and bottom edges will have a large complement of no data cells), but the defined pixels will have the same content (that is, a stitch() of both layers will be the same, modulo the no data cells on the boundary required for padding).

### CODE REVIEW

**Comment by metasim** [comment]
This is super helpful in RasterFrames too, where we're already seeing the need for user's to dynamically change the global tile size when rows are too coarse.

File: spark/src/main/scala/geotrellis/spark/regrid/Implicits.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.apache.spark.rdd.RDD
+
+import scala.reflect.ClassTag
+
+object Implicits extends Implicits
+
+trait Implicits {
+  implicit class withRegridMethods[
+    K: SpatialComponent: ClassTag,
+    V <: CellGrid: ClassTag: Stitcher: (? => CropMethods[V]): (? => TilePrototypeMethods[V]),
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/Implicits.scala]: `TilePrototypeMethods[V]` is not used

File: spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    // let the interval be start to end, inclusive
+    def intersect(that: Interval[N]) = {
+        Interval(if (ord.compare(this.start, that.start) > 0) this.start else that.start, 
+                 if (ord.compare(this.end, that.end) < 0) this.end else that.end)
+    }
+  }
+
+  def apply[
+    K: SpatialComponent: ClassTag,
+    V <: CellGrid: ClassTag: Stitcher: (? => CropMethods[V]): (? => TilePrototypeMethods[V]),
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala]: `TilePrototypeMethods[V]` is not used

File: spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+          .flatMap{ case (key, oldTile) => {
+            val SpatialKey(oldX, oldY) = key
+
+            val oldXstart = oldX.toLong * oldW.toLong
+            val oldYstart = oldY.toLong * oldH.toLong
+            val oldXrange = Interval(oldXstart, oldXstart + oldW - 1)
+            val oldYrange = Interval(oldYstart, oldYstart + oldH - 1)
+
+            val tileEx = ld.mapTransform(key)
+            val newBounds = nld.mapTransform(tileEx)
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala]: `tileEx` and `newBounds` are never used too

File: spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import geotrellis.vector._
+
+import scala.reflect._
+import scala.collection.mutable
+
+object Regrid {
+
+  private case class Interval[N : Ordering](start: N, end: N) {
+    val ord = implicitly[Ordering[N]]
+    // assert(ord.compare(start, end) < 1)
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala]: Do we need to uncomment this line?

File: spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+          .groupByKey
+          .mapValues { tiles => implicitly[Stitcher[V]].stitch(tiles, tileCols, tileRows) }
+
+      ContextRDD(tiled, newMd)
+    }
+  }
+
+  def apply[
+    K: SpatialComponent: ClassTag,
+    V <: CellGrid: ClassTag: Stitcher: (? => CropMethods[V]): (? => TilePrototypeMethods[V]),
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala]: `TilePrototypeMethods[V]` is not used

File: spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+import geotrellis.raster.crop._
+import geotrellis.raster.prototype._
+import geotrellis.raster.stitch._
+import geotrellis.spark._
+import geotrellis.spark.tiling._
+import geotrellis.util._
+import geotrellis.vector._
+
+import scala.reflect._
+import scala.collection.mutable
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala]: Unused import

File: spark/src/main/scala/geotrellis/spark/regrid/RegridMethods.scala [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import geotrellis.spark.tiling._
+import geotrellis.util._
+
+import org.apache.spark.rdd.RDD
+
+import scala.reflect.ClassTag
+
+class RegridMethods[
+  K: SpatialComponent: ClassTag,
+  V <: CellGrid: ClassTag: Stitcher: (? => CropMethods[V]): (? => TilePrototypeMethods[V]),
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/RegridMethods.scala]: `TilePrototypeMethods[V]` is not used

File: spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+            val oldXstart = oldX.toLong * oldW.toLong
+            val oldYstart = oldY.toLong * oldH.toLong
+            val oldXrange = Interval(oldXstart, oldXstart + oldW - 1)
+            val oldYrange = Interval(oldYstart, oldYstart + oldH - 1)
+
+            val tileEx = ld.mapTransform(key)
+            val newBounds = nld.mapTransform(tileEx)
+
+            for (
+              x <- (oldXstart / tileCols).toInt to (oldXrange.end.toDouble / tileCols).toInt ;
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala]: Hm interesting syntax with semicolons, but I like it

File: spark/src/test/scala/geotrellis/spark/regrid/Regrid.scala [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                                           LatLng,
+                                           KeyBounds[SpatialKey](SpatialKey(0,0), SpatialKey(3,2)))
+    ContextRDD(rdd, md)
+  }
+
+  describe("Regridding") {
+    it("should allow chipping into smaller tiles") {
+      val newLayer = simpleLayer.regrid(16)
+
+      // import geotrellis.raster.render._
```
Comment (pomadchin) [inline spark/src/test/scala/geotrellis/spark/regrid/Regrid.scala]: Mb tests cleanup / more tests are required here?

**Comment by pomadchin** [comment]
Btw I'm wondering how it would work on Spark, for instance in Zeppelin

File: spark/src/test/scala/geotrellis/spark/regrid/Regrid.scala (hunk 9 above)
Comment (jpolchlo) [reply spark/src/test/scala/geotrellis/spark/regrid/Regrid.scala]: Willing to take suggestions about tests that will cover cases that I don't.  These tests check what happens when we evenly break the source into smaller units, coalesce tiles into larger units, and when we extend in one axis and contract in the other (also covering the case when the target domain's extent is larger than the source domain).  I can't think of what I might be missing.

File: spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import geotrellis.util._
+import geotrellis.vector._
+
+import scala.reflect._
+
+object Regrid {
+
+  private case class Interval[N : Ordering](start: N, end: N) {
+    val ord = implicitly[Ordering[N]]
+    assert(ord.compare(start, end) < 1)
```
Comment (pomadchin) [inline spark/src/main/scala/geotrellis/spark/regrid/Regrid.scala]: Hm, mb to replace it with a require, to provide users some readble exception?

```scala
assert(ord.compare(start, end) < 1, "some friendly message")
```

**Pull request merged**